File: sliding_window/sliding_window1_lc28.py


# 定长的滑动窗口
def strStr(haystack, needle):
    """
        实现strStr()
    """
    start = 0
    for end in range(len(needle) - 1, len(haystack)):   # 这是一个定长的窗口(28和567都是定长)
        # 这种写法可以自动更新end, 我们只要考虑什么时候更新start就可以
        if haystack[start:end+1] == needle:
            return start
        start += 1              # 不相等只需要移动start, end会在循环的时候自动移动
    return -1

aaa = strStr('hello', 'll')
bbb = strStr('aaaaa', 'bba')
print(aaa)
print(bbb)

File: cpp/array/sliding_window/sliding_window1_lc28.cpp
/*
    找出字符串中第一个匹配项的下标
*/

#include <iostream>
using namespace std;

int strStr(string haystack, string needle)
{
    int start = 0;
    for (int end = needle.size() - 1; end < haystack.size(); end++)
    {
        if (haystack.substr(start, needle.size()) == needle)    // s.substr(pos, size)，取字符串s的子串，从pos位置开始，取size长度
        {
            return start;
        }
        start++;
    }
    return -1;
}


int main()
{
    string haystack = "hello";
    string needle = "ll";
    int result = strStr(haystack, needle);
    cout << result << endl; // 2
    return 0;
}